src/auth.rs

```diff
@@ -169,33 +200,26 @@ where
             .get::<Option<User>>()
             .expect("Auth extension missing. Is the auth layer installed?");
 
-        let expected_role = self.role.clone();
-        let actual_role = user.as_ref().and_then(|user| user.get_role());
+        let is_full = (RangeFull.start_bound(), RangeFull.end_bound()) == self.role_bounds.bounds();
 
-        match (expected_role, user, actual_role) {
-            (None, Some(user), None | Some(_)) => {
-                let user = user.clone();
-                request.extensions_mut().insert(user);
-
-                Ok(())
-            }
-            (Some(expected_role), Some(user), Some(actual_role))
-                if expected_role == actual_role =>
+        if let Some(user) = user {
+            if user
+                .get_role()
+                .map_or(is_full, |role| self.role_bounds.contains(&role))
```

Comment:
Is there a semantic difference between the full range and contains and if so do we want to use either to mean the same thing?

Reply:
There are two cases to take into account in this method:
1. Someone used the `login` method to declare the auth requirements or used the `login_with_role` method with a `..` (`full`) range (those two are semantically equal, since the `login` method uses the `full` range declaration).
2. Someone declared a specific range different from the `..` (`full`) range.

In the first case the `get_role` method can return `None`, if that happens we need to check if the `full` range was declared, then we have to just allow the user as-is, without checking the role completely. This completely covers case 1 and backwards compatibility, since the role system may be completely unused.

Case 2 is then handled by the `contains` method on `role_bounds`.

Have I missed something?

Reply:
Thanks for the detailed explanation!

I'm probably missing an important detail here, but [`contains` will return `true` for `None`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f01be4f706bd9d6191bdc6fd74a34291) and likewise the correct boolean when the `Some<T>` variant is returned.

Given that could we simply called `contains` over the value of `get_role`?

Reply:
With the other suggested change, I think we might be able to write this as something like:

```rust
        match user {
            Some(user) if self.role_bounds.contains(user.get_role()) => {
                let user = user.clone();
                request.extensions_mut().insert(user);

                Ok(())
            }

            _ => {
                let unauthorized_response = Response::builder()
                    .status(http::StatusCode::UNAUTHORIZED)
                    .body(Default::default())
                    .unwrap();

                Err(unauthorized_response)
            }
        }
```